src/utils_wmi.c

```diff
@@ -0,0 +1,427 @@
+                         VARIANT *result) {
+  DISPPARAMS params = {
+      .cArgs = 0, .cNamedArgs = 0,
+  };
+  return wmi_make_call(dispatcher, property_name, &params, result,
+                       DISPATCH_PROPERTYGET);
+}
+
+wmi_result_list_t *wmi_query(wmi_connection_t *connection, const char *query) {
+  HRESULT status = 0;
+  wmi_result_list_t *res = NULL;
+
+  VARIANT methodResult;
+  VARIANT propertyResult;
+  VARIANTARG args[1];
+  DISPPARAMS params;
+
+  params.cNamedArgs = 0;
+  params.rgvarg = args;
+
+  wchar_t *wquery = strtowstr(query);
+  params.cArgs = 1;
+  params.rgvarg[0].vt = VT_BSTR;
+  params.rgvarg[0].bstrVal = SysAllocString(wquery);
+  free(wquery);
+
+  status = wmi_invoke_method(connection->dispatcher, L"ExecQuery", &params,
+                             &methodResult);
+  SysFreeString(params.rgvarg[0].bstrVal);
+  VariantClear(args);
+  if (FAILED(status)) {
+    free(res);
+    VariantClear(&methodResult);
+    log_err("unknown error [0x%x] during query: '%s'. Error details: "
+            "ExecQuery() failed.",
+            (unsigned)status, query);
+    return NULL;
+  }
+
+  res = malloc(sizeof(wmi_result_list_t));
+  res->results = methodResult.pdispVal;
+  methodResult.pdispVal = NULL;
+  VariantClear(&methodResult);
+
+  status = wmi_get_property(res->results, L"Count", &propertyResult);
+  if (FAILED(status)) {
+    // It is very likely that the set returned by ExecQuery is empty,
+    // which seems to be signaled by a missing 'Count' property.
+    // Because of that, we're not going to treat it as an error.
+    res->count = 0;
+  } else {
+    res->count = propertyResult.intVal;
+  }
+  res->last_result = -1;
+  VariantClear(&propertyResult);
+
+  return res;
+}
+
+void wmi_result_list_release(wmi_result_list_t *results) {
+  if (results)
+    results->results->lpVtbl->Release(results->results);
+  free(results);
+}
+
+wmi_result_t *wmi_get_next_result(wmi_result_list_t *results) {
+  if (!results)
+    return NULL;
+
+  if (results->last_result + 1 >= results->count)
+    return NULL;
+
+  HRESULT hr;
+  VARIANTARG args[1] = {{
+      .vt = VT_UI4, .uintVal = results->last_result + 1,
+  }};
+  DISPPARAMS params = {
+      .cArgs = 1, .cNamedArgs = 0, .rgvarg = args,
+  };
+  VARIANT varResult;
+
+  hr = wmi_invoke_method(results->results, L"ItemIndex", &params, &varResult);
+  if (FAILED(hr)) {
+    VariantClear(&varResult);
+    VariantClear(args);
+    log_err("cannot get next result. Error code 0x%x", (unsigned)hr);
+    return NULL;
+  }
+
+  results->last_result++;
+
+  wmi_result_t *result = malloc(sizeof(wmi_result_t));
+  result->result = varResult.pdispVal;
+  varResult.pdispVal = NULL;
+  VariantClear(&varResult);
+  VariantClear(args);
+  return result;
+}
+
+void wmi_result_release(wmi_result_t *result) {
+  if (result)
+    result->result->lpVtbl->Release(result->result);
+  free(result);
+}
+
+void handle_error(HRESULT hr, const char *property_name) {

```

Comment:
Looks like this can be static. Also, log_error seems like a better name.

Reply:
Done. Changed to log_hresult_error so it's not confused with the log_err macro defined at the top.